include/counter.h

```diff
@@ -420,6 +447,71 @@ static inline u32_t z_impl_counter_get_max_relative_alarm(struct device *dev)
 	return api->get_max_relative_alarm(dev);
 }
 
+/**
+ * @brief Set guard period in ticks.
+ *
+ * Setting non-zero guard period enables detection of setting absolute alarm
+ * too late. It limits how far in the future absolute alarm can be set.
+ *
+ * Detection of too late setting is vital since if it is not detected alarm
+ * is delayed by full period of the counter (up to 32 bits). Because of the
+ * wrapping, it is impossible to distinguish alarm which is short in the past
+ * from alarm which is targeted to expire after full counter period. In order to
+ * detect too late setting, longest possible alarm is limited. Absolute value
+ * cannot exceed: (now + top_value - guard_period) % top_value.
+ *
+ * Guard period depends on application and counter frequency. If it is expected
+ * that absolute alarms setting might be delayed then guard period should
+ * exceed maximal potential delay. If use case allows, guard period can be set
+ * very high (e.g. half of the counter top value).
+ *
+ * After initialization guard period is set to 0 and late detection is disabled.
+ *
+ * @param dev		Pointer to the device structure for the driver instance.
+ * @param guard		Guard period in counter ticks.
+ * @param flags		Reserved for future use.
```

Comment:
We need to define and start using "late guard" flag already. I.e. There should be no need to change the application code if a new flag is added in the future. Also, maybe instead of `flags` we should use an enum. i.e.
```
counter_set_guard_period(struct device *dev, u32_t period, enum guard_type guard);
```

Reply:
Added `COUNTER_GUARD_PERIOD_LATE_ABS`.

i would prefer to keep flags. that allows further extensions, e.g. adding flag `COUNTER_GUARD_PERIOD_IN_US` to work in time domain.